Indexer-agent: Monitor submitted proofs of indexing for potential disputes

fordN opened the pull request:
This PR adds a step to the reconciliation loop of the indexer-agent to check on-chain proofs of indexing against their data to identify potential disputes.  

The agent looks for any closed allocations that are within the dispute period and checks the allocations POI against reference POIs generated from their graph node (if available); if the reference POIs do not match the allocation the allocation is flagged as a potential dispute and saved in the indexer database.  In addition, a `disputes get` command has been added to the indexer-cli for fetching the allocations that have been flagged as potential disputes.

fordN committed:
indexer-common: Add POI disputes to indexer management API

fordN committed:
indexer-common: Include tests for POI dispute resolvers

fordN committed:
indexer-agent: Monitor POIs, flag and store potential disputes

Jannis committed:
indexer-agent,indexer-common: Updated arbiration charter

- Include fixmes and todos for updated arbitration charter which
allows POI from previous epoch on allocation close.
- Improve error message suggestions.
- Update allocation.indexer type -> Address.
- address hexstring length should be 20.

sharppanda committed:
Update errors.md

Typo

fordN committed:
indexer-agent: Paginate when fetching disputable allocations

fordN committed:
indexer-agent: Verify POI against both closed epoch and prior epoch POIs

fordN committed:
indexer-agent: Include tests for POI dispute storing and fetching

fordN committed:
indexer-agent: Add closed and prior epoch data to POI disputes model

fordN committed:
indexer-agent, indexer-common: Improved log messages and formatting

fordN committed:
indexer-cli: Add `disputes get` command to fetch stored potential disputes

abarmat commented on packages/indexer-agent/src/agent.ts:
```diff
@@ -347,6 +369,105 @@ class Agent {
+            closedEpoch: allocation.closedAtEpoch,
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            closedEpochReferenceProof: rewardsPool!.referencePOI!,
+            closedEpochStartBlockHash: allocation.closedAtBlockHash,
```
Shouldn't this be `allocation.closedAtEpochStartBlockHash` ?

abarmat commented on packages/indexer-agent/src/network.ts:
```diff
@@ -435,28 +446,167 @@ export class Network {
+    try {
+      const zeroPOI = utils.hexlify(Array(32).fill(0))
+      // TODO: Use disputable epochs period value from the network instead of hardcoding to 2 epochs
+      const disputableEpoch = currentEpoch.toNumber() - 2
```
Interesting. We can set a constant for this, or even better to read from a config variable. Disputable allocations will depend very much on a social contract called the Arbitration Charter, probably we want let indexers to change this number easily.

fordN replied:
I've added a startup parameter to configure this constant, `--poi-disputable-epochs`.

abarmat commented on packages/indexer-agent/src/agent.ts:
```diff
@@ -256,6 +259,20 @@ class Agent {
 
+    const disputableAllocations = join({
+      currentEpoch,
+      activeDeployments,
```
This means we are checking disputable allocations only using the indexer active subgraph deployments, is that right?

fordN replied:
correct, the indexer will only be able to generate reference POIs for subgraphs they are actively indexing.

fordN committed:
indexer-agent: Add startup params to configure POI dispute monitoring

--poi-dispute-monitoring boolean to turn it on and off
--poi-disputable-epochs to configure how far back to look for close
allocations that could potentially be successfully disputed

abarmat reviewed (approved):
🚀
